atp/client.go

```diff
@@ -465,28 +520,34 @@ func (c *client) prepareResultChannels(
 	return nil
 }
 
-// getResultV2 works with the channels that communicate with the RuntimeMessage loop.
+// getResultV2 communicates with the RuntimeMessage loop to get the .
 func (c *client) getResultV2(
 	stepData schema.Input,
 ) ExecutionResult {
 	c.mutex.Lock()
-	resultChannel, found := c.runningStepResultChannels[stepData.RunID]
-	c.mutex.Unlock()
+	resultEntry, found := c.runningStepResultEntries[stepData.RunID]
+	c.logger.Debugf("Got result entry for run ID %q", stepData.RunID)
 	if !found {
 		return NewErrorExecutionResult(
-			fmt.Errorf("could not find result channel for step with run ID '%s'",
-				stepData.RunID),
+			fmt.Errorf("could not find result entry for step with run ID '%s'. Existing entries: %v",
+				stepData.RunID, c.runningStepResultEntries),
 		)
 	}
-	// Wait for the result
-	result, received := <-resultChannel
-	if !received {
+	if resultEntry.result == nil {
+		// Wait for the result
+		resultEntry.condition.Wait()
+	}
+	if resultEntry.result == nil {
 		return NewErrorExecutionResult(
-			fmt.Errorf("did not receive result from results channel in ATP client for step with run ID '%s'",
+			fmt.Errorf("did not receive result from results entry in ATP client for step with run ID '%s'",
 				stepData.RunID),
 		)
 	}
```

Comment:
This code is incorrect -- [the wait should be done inside a loop](https://pkg.go.dev/sync@go1.21.5#Cond.Wait):
```go
	for resultEntry.result == nil {
		resultEntry.condition.Wait()
	}
```
Thus, you don't need to check for the error condition at line 540, because it cannot happen.  (Also, omit the comment -- it adds nothing to the code.)

Reply:
That would make sense in the case of a shared condition variable, but with one, instead of getting stuck in a wait-loop that never exits, I'd rather that it fails. Unless you know of a case that would cause the condition variable to exit unprompted by the other section of code.

Reply:
It is possible that Go makes better guarantees than the typical implementation of Condition Variables, but normally CV waits are permitted to return "spuriously" (i.e., without an obvious signal or broadcast having been sent).  This means that you have to code the wait inside a "while loop" which checks the predicate and re-waits if it's inappropriate to proceed.  The result of this is better, more robust code...even if you never actually receive spurious wake-ups.  This code is halfway there, doing the wait under an `if` condition...if you change that to a `for`, then you're there.

The waiter is supposed to wait until the predicate condition is met -- the Condition Variable is a means toward that end, but it is not like an _Event_ -- the _state_ (i.e., the predicate) is **seperate** from the CV object embodied instead in the surrounding code.  Control returning from the CV wait is not significant in and of itself; and it should not constitute a failure if it does (which is why you do it in a loop).

Replacing the loop with an if-failure test doesn't actually solve any problems -- if the if-test fails, that means either that the signaler failed to set the predicate or it or some third party cleared the predicate again before the waiter could see it...neither of which has anything to do with the waiter or its code.  If the waiter waits forever (e.g., because it went around the loop and re-waited), again, that's not a bug on the wait-side of the code...and, the hang will indicate a problem just as the fail message would have.

Reply:
I checked the documentation, and it's **guaranteed** that in Go's implementation it will only return if `Signal()` or `Broadcast()` are called. The documentation recommends a loop if there is a chance that the condition that's being waited for could be changed when the lock is not acquired during the wait. However, that does not apply since the condition that is being waited for is terminal.
And because of that, I can also guarantee that if it ever needs to loop again, it will hang the goroutine. There are no possibilities where it will not hang it in that scenario.

Thank you for the wisdom, and I will keep it in mind when I use condition variables without Go's guarantees. But as designed, it will just hang and remove useful logs.